File: internal/fees/fees.go
package fees

import (
	"github.com/ochanomizu/predmarket-scanner/pkg/types"
)

var (
	FeeRateBPS     = 625
	ApplyFees      = true
)

func CalculatePolymarketFee(profit float64, market types.Market) float64 {
	if !ApplyFees || profit <= 0 {
		return 0
	}

	var avgPrice float64
	for _, outcome := range market.Outcomes {
		avgPrice += outcome.Price
	}
	if len(market.Outcomes) > 0 {
		avgPrice /= float64(len(market.Outcomes))
	}

	feeRate := float64(FeeRateBPS) / 10000.0
	effectiveFeeRate := feeRate * avgPrice * (1 - avgPrice)

	return profit * effectiveFeeRate
}

func CalculateNetROI(grossProfit, cost, fees float64) float64 {
	netPayout := cost + grossProfit - fees
	return (netPayout - cost) / cost
}

File: src/fees.rs
use crate::types::Market;

pub fn calculate_polymarket_fee(profit: f64, _market: &Market) -> f64 {
    const TRADING_FEE_RATE: f64 = 0.02;
    const MAKER_REBATE: f64 = 0.0002;

    let fee = profit * TRADING_FEE_RATE;
    let rebate = profit * MAKER_REBATE;

    fee - rebate
}

pub fn calculate_net_roi(gross_profit: f64, cost: f64, fees: f64) -> f64 {
    let net_payout = cost + gross_profit - fees;
    (net_payout - cost) / cost
}
